File: mobile/testopithecus-common/native/PAL/network/DefaultSyncNetwork.kt
package com.yandex.xplat.testopithecus.common

import com.yandex.xplat.common.*
import java.net.SocketTimeoutException
import java.net.URI
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okio.Buffer

val NetworkMethod.method: String
    get() = when (this) {
        NetworkMethod.get -> "GET"
        NetworkMethod.post -> "POST"
    }

class DefaultSyncNetwork(
    private val jsonSerializer: JSONSerializer,
    private val logger: Logger,
    networkConfig: NetworkConfig? = null
) : SyncNetwork {
    private val httpClient = OkHttpClient.Builder()
        .apply {
            networkConfig?.sslContextCreator?.createSSLConfiguredClient(this)
            networkConfig?.stethoProxy?.patch(this)
        }
        .build()

    override fun syncExecute(baseUrl: String, request: NetworkRequest, oauthToken: String?): Result<String> = runRequest(createHttpRequest(baseUrl, request, oauthToken))
    override fun syncExecuteWithRetries(
        retries: Int,
        baseUrl: String,
        request: NetworkRequest,
        oauthToken: String?
    ): Result<String> {
        var result: Result<String>? = null
        var retriesLeft = retries
        while (retriesLeft >= 0) {
            result = syncExecute(baseUrl, request, oauthToken)
            if (result!!.isValue()) {
                return result!!
            }
            retriesLeft--
        }
        return result!!
    }

    private fun runRequest(request: Request): Result<String> {
        logger.info(this.stringRepresentation(request))
        var response: Response? = null
        try {
            response = httpClient.newCall(request).execute()
        } catch (e: SocketTimeoutException) {
            return resultError(YSError("Request timeout!"))
        }
        if ((response == null) || (response.code() / 10 != 20)) {
            val responseBody = if (response == null) "" else response!!.body()?.string()
            val message = "Non 20x response code for request ${request.url()}, message: $responseBody"
            return resultError(YSError(message))
        }
        return resultValue(response.body()!!.string())
    }

    private fun stringRepresentation(request: Request): String {
        val sb = java.lang.StringBuilder()
        sb.append("curl ").append("'${request.url()}'").append(" -d ")

        val copy = request.newBuilder().build()
        val buffer = Buffer()
        copy.body()?.writeTo(buffer)
        val data = buffer.readUtf8()

        sb.append("'$data'")
        for ((key, value) in request.headers().toMultimap()) {
            sb.append(" -H '$key: ${value[0]}'")
        }
        return sb.toString()
    }

    private fun createHttpRequest(baseUrl: String, request: NetworkRequest, oauthToken: String?): Request {
        val encodedParameters = encodeRequest(this.jsonSerializer, request.encoding(), request.method(), request.params())

        val urlBuilder = HttpUrl.get(URI.create(baseUrl))!!.newBuilder()
        urlBuilder.addPathSegments(request.targetPath())

        val queryParameters: Map<String, *> = JSONItemSupport.from(request.urlExtra()) as Map<String, *> + encodedParameters.queryParameters
        queryParameters.forEach { (key, value) ->
            stringifyQueryParam(value)?.let { urlBuilder.addQueryParameter(key, it) }
        }

        val builder = Request.Builder().url(urlBuilder.build())
            .addHeader("Connection", "keep-alive")
            .addHeader("Content-Type", "application/json")
        if (oauthToken != null) {
            builder.addHeader("Authorization", "OAuth $oauthToken")
        }

        val headers = JSONItemSupport.from(request.headersExtra()) as Map<String, *>
        headers.forEach { (key, value) ->
            stringifyQueryParam(value)?.let { builder.addHeader(key, it) }
        }

        if (request.params().asMap().isEmpty()) {
            builder.method(request.method().method, null)
        } else {
            builder.method(request.method().method, encodedParameters.body)
        }
        return builder.build()
    }
}

fun stringifyQueryParam(value: Any?): String? =
    when (value) {
        is Number -> value.toString()
        is String -> value
        true -> "yes"
        false -> "no"
        null -> "null"
        is StringJSONItem -> value.value
        is IntegerJSONItem -> value.asInt64().toString()
        is BooleanJSONItem -> value.value.toString()
        else -> throw NotValidQueryTypeException()
    }

class NotValidQueryTypeException : Exception()

File: mobile/testopithecus-common/native/PAL/network/DefaultSyncNetwork.swift
//
//  DefaultNetworkFetch.swift
//  XMail
//
//  Created by Dmitry Zakharov.
//  Copyright © 2018 Yandex. All rights reserved.
//

import Foundation

public final class DefaultSyncNetwork: NSObject, SyncNetwork {
  private let logger: Logger
  private let errorThrower: ErrorThrower
  private let urlSession: URLSession

  private lazy var decoder: JSONDecoder = JSONDecoder()

  public override init() {
    logger = Log.getDefaultLogger()
    errorThrower = Registry.get().errorThrower
    urlSession = URLSession(configuration: .default, delegate: DefaultURLSessionDelegate(), delegateQueue: nil)
  }

  public func syncExecute(_ baseUrl: String, _ request: NetworkRequest, _ oauthToken: String!) -> Result<String> {
    guard let httpRequest = buildRequest(baseUrl: baseUrl, request: request, token: oauthToken) else {
      errorThrower.fail("Can't build request!")
      fatalError("Can't build request!")
    }
    logger.info("Performing: \(httpRequest.url?.absoluteString ?? "")")
    let semaphore = DispatchSemaphore(value: 0)
    var result: String?
    var error: String = ""
    var responseStatusCode: Int = 200
    let task = urlSession.dataTask(with: httpRequest) { data, response, errorData in
      if let data = data {
        result = String(data: data, encoding: .utf8)
      }
      if let response = response as? HTTPURLResponse {
        responseStatusCode = response.statusCode
      }
      if let errorData = errorData {
        error = errorData.localizedDescription
      }
      semaphore.signal()
    }
    task.resume()
    semaphore.wait()
    if result == nil {
      return resultError(YSError("Http request failure! " + error))
    }
    if request.targetPath().startsWith("v2"),
       responseStatusCode >= 400 {
      return resultError(YSError("Http response status code: \(responseStatusCode)"))
    }
    return resultValue(result!)
  }

  public func syncExecuteWithRetries(_ retries: Int32, _ baseUrl: String, _ request: NetworkRequest, _ oauthToken: String!) -> Result<String> {
    var retriesLeft: Int32 = retries
    var result: Result<String>?
    while retriesLeft >= 0 {
      result = syncExecute(baseUrl, request, oauthToken)
      if result!.isValue() {
        return result!
      }
      retriesLeft -= 1
    }
    return result!
  }

  private func buildRequest(baseUrl: String, request: NetworkRequest, token: String?) -> URLRequest? {
    guard let url = url(for: request, baseUrl: baseUrl) else {
      return nil
    }
    var httpRequest = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: TimeInterval(60))
    if let token = token {
      httpRequest.setValue("OAuth \(token)", forHTTPHeaderField: "Authorization")
    }
    request.headersExtra().asMap().items.forEach { key, value in
      if let param = stringifyQueryParam(value) {
        httpRequest.setValue(param, forHTTPHeaderField: key)
      }
    }
    httpRequest.httpMethod = request.method().toString()
    return encodeRequest(httpRequest, with: request.encoding(), params: request.params())
  }

  private func stringifyQueryParam(_ value: JSONItem) -> String? {
    switch value.kind {
    case .double:
      return doubleToString((value as! DoubleJSONItem).value)
    case .integer:
      return int64ToString((value as! IntegerJSONItem).asInt64())
    case .boolean:
      return (value as! BooleanJSONItem).value ? "yes" : "no"
    case .string:
      return (value as! StringJSONItem).value
    case .nullItem:
      return "null"
    case .map:
      fatalError("Maps are not supported as URL query request parameters: \(String(describing: value))")
    case .array:
      fatalError("Arrays are not supported as URL query request parameters: \(String(describing: value))")
    }
    return nil
  }

  private func url(for request: NetworkRequest, baseUrl: String) -> URL? {
    let url = fullPath(for: request, baseUrl: baseUrl)
    guard var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: true) else {
      NSLog("Unable to build URL for \(request.method()) request; P: \(request.targetPath())")
      return nil
    }
    let items = queryItems(fromUrlExtra: request.urlExtra())
    urlComponents.queryItems = [URLQueryItem(name: "client", value: "iphone")]
    if !items.isEmpty {
      urlComponents.queryItems?.append(contentsOf: items)
    }

    return urlComponents.url
  }

  private func fullPath(for request: NetworkRequest, baseUrl: String) -> URL {
    let baseURL = URL(string: baseUrl)!
    return baseURL.appendingPathComponent(request.targetPath())
  }
}
